src/Columns/ColumnReplicated.cpp

```diff
@@ -0,0 +1,678 @@
+{
+    return nested_column->getFloat64(indexes.getIndexAt(n));
+}
+
+Float32 ColumnReplicated::getFloat32(size_t n) const
+{
+    return nested_column->getFloat32(indexes.getIndexAt(n));
+}
+
+UInt64 ColumnReplicated::getUInt(size_t n) const
+{
+    return nested_column->getUInt(indexes.getIndexAt(n));
+}
+
+Int64 ColumnReplicated::getInt(size_t n) const
+{
+    return nested_column->getInt(indexes.getIndexAt(n));
+}
+
+UInt64 ColumnReplicated::get64(size_t n) const
+{
+    return nested_column->get64(indexes.getIndexAt(n));
+}
+
+StringRef ColumnReplicated::getDataAt(size_t n) const
+{
+    return nested_column->getDataAt(indexes.getIndexAt(n));
+}
+
+ColumnPtr ColumnReplicated::convertToFullColumnIfReplicated() const
+{
+    return nested_column->index(*indexes.getIndexes(), 0);
+}
+
+void ColumnReplicated::insertData(const char * pos, size_t length)
+{
+    nested_column->insertData(pos, length);
+    indexes.insertIndex(nested_column->size() - 1);
+}
+
+StringRef ColumnReplicated::serializeValueIntoArena(size_t n, Arena & arena, char const *& begin) const
+{
+    return nested_column->serializeValueIntoArena(indexes.getIndexAt(n), arena, begin);
+}
+
+StringRef ColumnReplicated::serializeAggregationStateValueIntoArena(size_t n, Arena & arena, char const *& begin) const
+{
+    return nested_column->serializeAggregationStateValueIntoArena(indexes.getIndexAt(n), arena, begin);
+}
+
+char * ColumnReplicated::serializeValueIntoMemory(size_t n, char * memory) const
+{
+    return nested_column->serializeValueIntoMemory(indexes.getIndexAt(n), memory);
+}
+
+std::optional<size_t> ColumnReplicated::getSerializedValueSize(size_t n) const
+{
+    return nested_column->getSerializedValueSize(indexes.getIndexAt(n));
+}
+
+const char * ColumnReplicated::deserializeAndInsertFromArena(const char * pos)
+{
+    const auto * res = nested_column->deserializeAndInsertFromArena(pos);
+    indexes.insertIndex(nested_column->size() - 1);
+    return res;
+}
+
+const char * ColumnReplicated::deserializeAndInsertAggregationStateValueFromArena(const char * pos)
+{
+    const auto * res = nested_column->deserializeAndInsertAggregationStateValueFromArena(pos);
+    indexes.insertIndex(nested_column->size() - 1);
+    return res;
+}
+
+const char * ColumnReplicated::skipSerializedInArena(const char * pos) const
+{
+    return nested_column->skipSerializedInArena(pos);
+}
+
+#if !defined(DEBUG_OR_SANITIZER_BUILD)
+void ColumnReplicated::insertRangeFrom(const IColumn & src, size_t start, size_t length)
+#else
+void ColumnReplicated::doInsertRangeFrom(const IColumn & src, size_t start, size_t length)
+#endif
+{
+    if (length == 0)
+        return;
+
+    if (start + length > src.size())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Parameter out of bound in ColumnReplicated::insertRangeFrom method.");
+
+    if (const auto * src_replicated = typeid_cast<const ColumnReplicated *>(&src))
+    {
+        auto & indexes_match = insertion_cache[src_replicated];
```

Comment:
This cache looks scary. There can be an ABA on src.

Do we call insertRangeFrom for ColumnReplicated from somewhere? I expect this column is constructed manually always, so maybe we can throw an exception for all the insertion attempts, initially?

Reply:
Stupid idea: introduce a field with a unique random id in `ColumnReplicated` and use it as a key for hash table instead of pointer